### PR TITLE
Add ResendResetPassword action to break up ResetPassword

### PR DESCRIPTION
**Pull request by bondydaa**
### Details

We were re-using an api action in two different places when we wanted different behavior so this splits them so that we can do different things when `Forgot?` is clicked vs `Resend Link` is clicked.

### Fixed Issues

$ https://github.com/Expensify/App/issues/10785

### Tests

- [x] Verify that no errors appear in the JS console
    - This is reported here https://expensify.slack.com/archives/C01GTK53T8Q/p1663264353337809?thread_ts=1662053176.714619&cid=C01GTK53T8Q and I bumped it.

#### Non SAML account test
1. Enter an email for an existing, non-saml account and click continue
1. Click the `Forgot?` button when prompted for you password.
1. Ensure that the new page that appears doesn't have a loader spinner appear and that no Brick Road messages appear. image
1. Click the `Resend link` button and ensure the success Brick Road message shows image

#### SAML account test
1. Enter an email that is on a SAML required domain (if you don't have this you can test with expensifail.com but ensure you make QA aware you're doing this on staging/production) and click continue
1. Click the `Forgot?` button when prompted for you password ensure the next page doesn't show a loader or any Brick Road messages
1. Ensure you get the proper error Brick Road message like so Screen Shot 2022-09-15 at 11 04 38 AM

#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

<details>
<summary><h4>PR Reviewer Checklist</h4>

The Contributor+ will copy/paste it into a new comment and complete it after the author checklist is completed
</summary>

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

</details>

### QA Steps

- [ ] Verify that no errors appear in the JS console

#### Non SAML account test
1. Enter an email for an existing, non-saml account and click continue
1. Click the `Forgot?` button when prompted for you password.
1. Ensure that the new page that appears doesn't have a loader spinner appear and that no Brick Road messages appear. image
1. Click the `Resend link` button and ensure the success Brick Road message shows image

#### SAML account test
1. Enter an email that is on a SAML required domain (if you don't have this you can test with expensifail.com but ensure you make QA aware you're doing this on staging/production) and click continue
1. Click the `Forgot?` button when prompted for you password ensure the next page doesn't show a loader or any Brick Road messages
1. Ensure you get the proper error Brick Road message like so Screen Shot 2022-09-15 at 11 04 38 AM

#### Desktop

No message or loader shows when you first click `Forgot?`
image

Message appears properly when you click resend
image

Error message is properly shown for SAML users
Screen Shot 2022-09-15 at 11 04 38 AM

### CODE REVIEW

File: src/libs/actions/Session/index.js [hunk 1]
```diff
@@ -289,6 +289,46 @@ function resetPassword() {
...
+                },
+            },
+        ],
+        successData: [
+            {
+                onyxMethod: CONST.ONYX.METHOD.MERGE,
+                key: ONYXKEYS.ACCOUNT,
+                value: {
+                    isLoading: false,
+                    message: null,
```
Comment (luacmartins) [inline src/libs/actions/Session/index.js:311]: NAB I'm wondering if we actually need to set `message: null` here. We already clear `message` in `optimisticData` so it should already be empty. This applies for `failureData` and for `resendResetPassword` too.

File: src/libs/actions/Session/index.js [hunk 2]
```diff
@@ -289,6 +289,46 @@ function resetPassword() {
     API.write('RequestPasswordReset', {
         email: credentials.login,
     },
+    {
+        optimisticData: [
+            {
+                onyxMethod: CONST.ONYX.METHOD.MERGE,
+                key: ONYXKEYS.ACCOUNT,
+                value: {
+                    isLoading: true,
```
Comment (luacmartins) [inline src/libs/actions/Session/index.js:298]: NAB maybe this is just me, but it looks a bit odd to be redirected to the Forgot page and see the loading indicator when the message above says we sent a link already. Should we just remove the loading indicator from the first load? 🤔

Comment (bondydaa) [reply src/libs/actions/Session/index.js:298]: Yep agreed.

File: src/libs/actions/Session/index.js (hunk 1 above)
Comment (bondydaa) [reply src/libs/actions/Session/index.js:311]: Yeah my only thought was if something else would have set this `message` key previously?

Though I just tested this and don't think it's possible, here's what I did
1. sign up with a new account, click resend to resend a 2nd validation link and get the success message which also uses `account.message`.
2. validate the account outside of going through new dot (I simulated this with cli tools to validate for me)
3. go back to "start" page and enter email again and then click `forgot?`. Here is where I was expecting the `message` previous to persist but it didn't so I think we're okay. There is probably a corner case somewhere lurking if we don't probably clear out `account.message` but we can fix that if it ever shows up easily.

**Comment by bondydaa** [comment]
Updated and ready for review. 

Though there is one thing I'm not sure about, we didn't want the Brick Road messaging showing up when the page first loaded but in the case for SAML required users it actually is and I think that's okay since it then gives a heads up that we didn't actually send the email.

File: src/libs/actions/Session/index.js (hunk 1 above)
Comment (luacmartins) [reply src/libs/actions/Session/index.js]: Whenever we tap `Forgot?` we trigger [resetPassword](https://github.com/Expensify/App/blob/b9e3ee8f76fcc0b69aae5c30dc8ba24ff9e00452/src/pages/signin/PasswordForm.js#L154) and we clear `message` and `error` in [optimisticData](https://github.com/Expensify/App/blob/b9e3ee8f76fcc0b69aae5c30dc8ba24ff9e00452/src/libs/actions/Session/index.js#L300-L301), so I think that we should be ok.

**Comment by luacmartins** [comment]
I think the SAML case makes sense.

File: src/libs/actions/Session/index.js [hunk 3]
```diff
@@ -289,6 +289,43 @@ function resetPassword() {
...
+        ],
+        failureData: [
+            {
+                onyxMethod: CONST.ONYX.METHOD.MERGE,
+                key: ONYXKEYS.ACCOUNT,
+                value: {
+                    isLoading: false,
+                },
+            },
+        ],
```
Comment (luacmartins) [inline src/libs/actions/Session/index.js:321]: Hmm now that we don't set `isLoading` in optimisticData, there's no need to set it to false here. So we can just get rid of both `successData` and `failureData`

**Review by luacmartins — DISMISSED** [review]

**Review by luacmartins — APPROVED** [review]

**Comment by mananjadhav** [comment]
Do I need any additional setup here? I am getting 404 errors in the network tab on clicking Resend Link

image

**Comment by luacmartins** [comment]
@mananjadhav the API changes that introduce this command are only in staging at the moment. You can either wait until these changes make it to prod (probably on Monday) or point your App to the staging API to test.

**Comment by mananjadhav** [comment]
> the API changes that introduce this command are only in staging

@luacmartins I thought so, I copied `.env.staging` into a `.env` but it gives me CORS error. I used to use a Chrome web plugin for CORS but that doesn't seem to work here. Do I need to do something additional for staging APIs to work?

**Comment by luacmartins** [comment]
@mananjadhav web might be a problem, but I don't have issues on desktop or mobile. Maybe you can test those?

**Comment by mananjadhav** [comment]
Yeah App and Desktop I'll sort out. Then I'll test on these platforms and Monday if it is pushed I'll test the Web too.

**Comment by mananjadhav** [comment]
Changes look good and I tested them on Desktop, iOS and Android App.  The only corner case that might not be related to the PR is, if I press `Resend Link` and immediately I go offline (before the request completes), the button gets disabled and is in loading state.

🎀 👀 🎀 
C+ reviewed

<details>
<summary><h4>PR Reviewer Checklist</h4>

The Contributor+ will copy/paste it into a new comment and complete it after the author checklist is completed
</summary>

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

</details>

**Comment by luacmartins** [comment]
@mananjadhav thanks for the review! I think that edge case exists for all commands that set a loading state, so I wouldn't block this PR on that. That being said, we should probably address the general issue since it could happen quite often with spotty connections.

**Comment by luacmartins** [comment]
Web-E PR is in prod. Removing hold and merging!

**Comment by luacmartins** [comment]
I edited the checklists so that tests pass.

**Pull request merged**